File: java/dbos-starter/src/main/java/com/example/dbosstarter/controller/DurableStarterController.java
package com.example.dbosstarter.controller;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RestController;

import com.example.dbosstarter.service.DurableStarterService;

import dev.dbos.transact.DBOS;
import dev.dbos.transact.StartWorkflowOptions;

import static com.example.dbosstarter.service.DurableStarterService.STEPS_EVENT;

@RestController
@CrossOrigin(origins = "*")
public class DurableStarterController {

    private static final Logger logger = LoggerFactory.getLogger(DurableStarterController.class);

    @Autowired
    private DurableStarterService service;

    @Autowired
    private DBOS dbos;

    @GetMapping("/workflow/{taskId}")
    public ResponseEntity<Void> startWorkflow(@PathVariable String taskId) {
        dbos.startWorkflow(() -> service.exampleWorkflow(), new StartWorkflowOptions(taskId));
        return ResponseEntity.ok().build();
    }

    @GetMapping("/last_step/{taskId}")
    public ResponseEntity<String> lastStep(@PathVariable String taskId) {
        var step = (Integer) dbos.getEvent(taskId, STEPS_EVENT, 0.0f);
        return ResponseEntity.ok(String.valueOf(step != null ? step : 0));
    }

    @PostMapping("/crash")
    public ResponseEntity<Void> crash() {
        logger.warn("Crash endpoint called - terminating application");
        Runtime.getRuntime().halt(0);
        return ResponseEntity.ok().build();
    }
}

File: kotlin/dbos-starter/src/main/kotlin/com/example/dbos_starter/controller/DurableStarterController.kt
package com.example.dbos_starter.controller

import com.example.dbos_starter.service.DurableStarterService
import com.example.dbos_starter.service.STEPS_EVENT
import dev.dbos.transact.DBOS
import dev.dbos.transact.StartWorkflowOptions
import dev.dbos.transact.execution.ThrowingRunnable
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.time.Duration
import dev.dbos.transact.startWorkflow

private val logger = LoggerFactory.getLogger(DurableStarterController::class.java)

@RestController
@CrossOrigin(origins = ["*"])
class DurableStarterController(private val service: DurableStarterService) {

    @GetMapping("/workflow/{taskId}")
    fun runWorkflow(@PathVariable taskId: String): ResponseEntity<Void> {
        startWorkflow(taskId) { service.exampleWorkflow() }
        return ResponseEntity.ok().build()
    }

    @GetMapping("/last_step/{taskId}")
    fun lastStep(@PathVariable taskId: String): ResponseEntity<String> {
        val step = DBOS.getEvent(taskId, STEPS_EVENT, Duration.ofSeconds(0)) as Int?
        return ResponseEntity.ok((step ?: 0).toString())
    }

    @PostMapping("/crash")
    fun crash(): ResponseEntity<Void> {
        logger.warn("Crash endpoint called - terminating application")
        Runtime.getRuntime().halt(0)
        return ResponseEntity.ok().build()
    }
}